libbindgen/src/ir/item.rs

```diff
@@ -688,23 +688,34 @@ impl Item {
             return base_name;
         }
 
-        if within_namespace {
-            return ctx.rust_mangle(&base_name).into_owned();
+        // Concatenate this item's ancestors' names together.
+        let mut names = vec![];
+        if !base_name.is_empty() {
+            names.push(base_name);
+        }
+
+        let mut current_ancestor_id = target.parent_id();
+
+        while current_ancestor_id != ctx.root_module() {
```

Comment:
I really prefer the `ancestors()` iterator ;) I think it is both more clear and more concise.

What do you think of this?

```rust
let mut names: Vec<_> = target.parent_id()
    .ancestors(ctx)

    //////// THIS IS THE NEW PART

    .take_while(|id| {
        if !within_namespace {
            true
        } else {
            // Stop iterating ancestors once we reach a namespace.
            !ctx.resolve_item(id).is_module()
        }
    })

    //////// END NEW PART

    .filter(|id| *id != ctx.root_module())
    .map(|id| {
        let item = ctx.resolve_item(id);
        let target = ctx.resolve_item(item.name_target(ctx, false));
        target.base_name(ctx, false)
    })
    .filter(|name| !name.is_empty())
    .collect();
```